File: go/json5_test.go
package json5

import (
	"math"
	"reflect"
	"testing"

	jsonic "github.com/jsonicjs/jsonic/go"
)

// parser returns a fresh Jsonic instance with the Json5 plugin installed
// with the given option overrides.
func parser(t *testing.T, overrides ...map[string]any) *jsonic.Jsonic {
	t.Helper()
	j := jsonic.Make()
	if err := j.UseDefaults(Json5, Defaults(), overrides...); err != nil {
		t.Fatalf("UseDefaults: %v", err)
	}
	return j
}

func parse(t *testing.T, j *jsonic.Jsonic, src string) any {
	t.Helper()
	v, err := j.Parse(src)
	if err != nil {
		t.Fatalf("Parse(%q) error: %v", src, err)
	}
	return v
}

func eq(t *testing.T, got, want any, src string) {
	t.Helper()
	if fg, ok := got.(float64); ok {
		if fw, ok := want.(float64); ok && math.IsNaN(fg) && math.IsNaN(fw) {
			return
		}
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("Parse(%q) = %#v, want %#v", src, got, want)
	}
}

func TestPrimitives(t *testing.T) {
	j := parser(t)
	cases := []struct {
		src  string
		want any
	}{
		{"true", true},
		{"false", false},
		{"null", nil},
		{`"hello"`, "hello"},
		{`'hello'`, "hello"},
		{"42", float64(42)},
		{"3.14", 3.14},
		{"-7", float64(-7)},
		{"+5", float64(5)},
		{".5", 0.5},
		{"5.", float64(5)},
		{"1e10", 1e10},
		{"1.5e-2", 0.015},
		{"0x1F", float64(31)},
		{"0xDEADBEEF", float64(0xDEADBEEF)},
		{"-0x10", float64(-16)},
	}
	for _, c := range cases {
		eq(t, parse(t, j, c.src), c.want, c.src)
	}
}

func TestInfinityNaN(t *testing.T) {
	j := parser(t)
	for _, src := range []string{"Infinity", "+Infinity"} {
		eq(t, parse(t, j, src), math.Inf(1), src)
	}
	eq(t, parse(t, j, "-Infinity"), math.Inf(-1), "-Infinity")
	for _, src := range []string{"NaN", "+NaN", "-NaN"} {
		eq(t, parse(t, j, src), math.NaN(), src)
	}

	// Can be disabled.
	jn := parser(t, map[string]any{"infinity": false})
	if _, err := jn.Parse("Infinity"); err == nil {
		t.Error("expected error for Infinity when infinity=false")
	}
}

func TestObjects(t *testing.T) {
	j := parser(t)
	cases := []struct {
		src  string
		want any
	}{
		{`{}`, map[string]any{}},
		{`{"a":1}`, map[string]any{"a": float64(1)}},
		{`{a:1}`, map[string]any{"a": float64(1)}},
		{`{a:1,b:2}`, map[string]any{"a": float64(1), "b": float64(2)}},
		{`{a:1, 'b':2}`, map[string]any{"a": float64(1), "b": float64(2)}},
		{`{ nested: { x: 1 } }`, map[string]any{"nested": map[string]any{"x": float64(1)}}},
		{`{$id:1, _n:2, a1:3}`, map[string]any{"$id": float64(1), "_n": float64(2), "a1": float64(3)}},
	}
	for _, c := range cases {
		eq(t, parse(t, j, c.src), c.want, c.src)
	}
}

func TestArrays(t *testing.T) {
	j := parser(t)
	cases := []struct {
		src  string
		want any
	}{
		{`[]`, []any{}},
		{`[1]`, []any{float64(1)}},
		{`[1,2,3]`, []any{float64(1), float64(2), float64(3)}},
		{`["a","b"]`, []any{"a", "b"}},
		{`[[1,2],[3,4]]`, []any{[]any{float64(1), float64(2)}, []any{float64(3), float64(4)}}},
	}
	for _, c := range cases {
		eq(t, parse(t, j, c.src), c.want, c.src)
	}
}

func TestTrailingCommas(t *testing.T) {
	j := parser(t)
	cases := []string{`[1,2,3,]`, `{a:1,b:2,}`, `[1,]`, `{a:1,}`, `[ 1 , 2 , ]`}
	for _, src := range cases {
		if _, err := j.Parse(src); err != nil {
			t.Errorf("Parse(%q) error: %v", src, err)
		}
	}
}

func TestComments(t *testing.T) {
	j := parser(t)
	cases := []string{
		"// hello\n42",
		"/* block */ 42",
		"{ a: 1, /* mid */ b: 2 }",
		"[/* a */ 1, /* b */ 2]",
		"/* multi\nline\ncomment */ [1,2]",
	}
	for _, src := range cases {
		if _, err := j.Parse(src); err != nil {
			t.Errorf("Parse(%q) error: %v", src, err)
		}
	}

	// Hash comments are not JSON5 by default.
	if _, err := j.Parse("# nope\n42"); err == nil {
		t.Error("expected error for hash comment")
	}

	// Can be enabled.
	jh := parser(t, map[string]any{"hashComment": true})
	eq(t, parse(t, jh, "# hello\n42"), float64(42), "# hello\\n42")
}

func TestStrings(t *testing.T) {
	j := parser(t)
	cases := []struct {
		src  string
		want string
	}{
		{`"hello"`, "hello"},
		{`'hello'`, "hello"},
		{`"he said \"hi\""`, `he said "hi"`},
		{`'he said \'hi\''`, `he said 'hi'`},
		{`"a\tb"`, "a\tb"},
		{`"a\nb"`, "a\nb"},
		{`"a\u0041b"`, "aAb"},
		{`"a\x41b"`, "aAb"},
		{`"\0"`, "\x00"},
		{"\"line1\\\nline2\"", "line1line2"},
	}
	for _, c := range cases {
		got := parse(t, j, c.src)
		if got != c.want {
			t.Errorf("Parse(%q) = %q, want %q", c.src, got, c.want)
		}
	}

	// Backticks not JSON5 by default.
	if _, err := j.Parse("`backtick`"); err == nil {
		t.Error("expected error for backtick string")
	}

	// Can be enabled.
	jb := parser(t, map[string]any{"backtickString": true})
	eq(t, parse(t, jb, "`backtick`"), "backtick", "`backtick`")
}

func TestRejectsNonJSON5(t *testing.T) {
	j := parser(t)
	cases := []string{
		"",
		"foo",
		"0o17",
		"0b101",
		"1_000",
		"1,2,3",
		"a:1",
	}
	for _, src := range cases {
		if v, err := j.Parse(src); err == nil {
			t.Errorf("Parse(%q) expected error, got %#v", src, v)
		}
	}
}

func TestNonStrictOptions(t *testing.T) {
	js := parser(t, map[string]any{
		"octal":           true,
		"binary":          true,
		"numberSeparator": true,
	})
	eq(t, parse(t, js, "0o17"), float64(15), "0o17")
	eq(t, parse(t, js, "0b101"), float64(5), "0b101")
	eq(t, parse(t, js, "1_000"), float64(1000), "1_000")

	jnh := parser(t, map[string]any{"hex": false})
	if _, err := jnh.Parse("0x1F"); err == nil {
		t.Error("expected error when hex=false")
	}
}

func TestStrictValueToggle(t *testing.T) {
	// With strictValue disabled, bare words parse as text strings
	// (Jsonic's default text fallback).
	j := parser(t, map[string]any{"strictValue": false})
	eq(t, parse(t, j, "foo"), "foo", "foo")
}

func TestJSON5SpecExample(t *testing.T) {
	j := parser(t)
	src := `{
      // comments
      unquoted: 'and you can quote me on that',
      singleQuotes: 'I can use "double quotes" here',
      lineBreaks: "Look, Mom! \
No \\n's!",
      hexadecimal: 0xdecaf,
      leadingDecimalPoint: .8675309, andTrailing: 8675309.,
      positiveSign: +1,
      trailingComma: 'in objects', andIn: ['arrays',],
      "backwardsCompatible": "with JSON",
    }`
	got := parse(t, j, src)
	want := map[string]any{
		"unquoted":            "and you can quote me on that",
		"singleQuotes":        `I can use "double quotes" here`,
		"lineBreaks":          "Look, Mom! No \\n's!",
		"hexadecimal":         float64(0xdecaf),
		"leadingDecimalPoint": 0.8675309,
		"andTrailing":         float64(8675309),
		"positiveSign":        float64(1),
		"trailingComma":       "in objects",
		"andIn":               []any{"arrays"},
		"backwardsCompatible": "with JSON",
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("spec example mismatch:\n got: %#v\nwant: %#v", got, want)
	}
}

func TestJSONIsJSON5(t *testing.T) {
	j := parser(t)
	cases := []string{
		`{}`,
		`[]`,
		`{"a":1,"b":"two","c":null,"d":true,"e":false}`,
		`[1,2.5,-3,1e10,"s",null,true,false]`,
		`{"nested":{"list":[1,{"x":null}]}}`,
	}
	for _, src := range cases {
		if _, err := j.Parse(src); err != nil {
			t.Errorf("Parse(%q) error: %v", src, err)
		}
	}
}

File: rs/tests/json5_test.rs
// In-language tests: the behaviour the shared fixtures cannot pin, and
// the same cases `go/json5_test.go` and `ts/test/json5.test.ts` carry,
// so the three suites can be read side by side.

mod common;

use std::sync::Arc;
use std::thread;

use tabnas::Tabnas;
use tabnas_json5::{json5, make, make_with, parse, parse_with, plugin, Json5Options};

use common::{json, to_value};

/// A fresh parser with the given overrides on the defaults.
fn parser(configure: impl FnOnce(&mut Json5Options)) -> Tabnas {
    let mut options = Json5Options::default();
    configure(&mut options);
    make_with(options)
}

fn parsed(parser: &Tabnas, src: &str) -> String {
    json(&parse_with(parser, src).unwrap_or_else(|error| panic!("{src:?}: {error}")))
}

fn code(parser: &Tabnas, src: &str) -> String {
    match parse_with(parser, src) {
        Ok(value) => panic!("{src:?} parsed to {}", json(&value)),
        Err(error) => error.code,
    }
}

// --- go/json5_test.go -------------------------------------------------

#[test]
fn primitives() {
    let j = make();
    for (src, want) in [
        ("true", "true"),
        ("false", "false"),
        ("null", "null"),
        (r#""hello""#, r#""hello""#),
        ("'hello'", r#""hello""#),
        ("42", "42"),
        ("3.14", "3.14"),
        ("-7", "-7"),
        ("+5", "5"),
        (".5", "0.5"),
        ("5.", "5"),
        ("1e10", "10000000000"),
        ("1.5e-2", "0.015"),
        ("0x1F", "31"),
        ("0xDEADBEEF", "3735928559"),
        ("-0x10", "-16"),
    ] {
        assert_eq!(parsed(&j, src), want, "{src:?}");
    }
}

#[test]
fn infinity_and_nan() {
    let j = make();
    for src in ["Infinity", "+Infinity"] {
        assert_eq!(parsed(&j, src), "Infinity", "{src:?}");
    }
    assert_eq!(parsed(&j, "-Infinity"), "-Infinity");
    for src in ["NaN", "+NaN", "-NaN"] {
        assert_eq!(parsed(&j, src), "NaN", "{src:?}");
    }

    // Can be disabled.
    let jn = parser(|o| o.infinity = false);
    assert_eq!(code(&jn, "Infinity"), "unexpected");
    assert_eq!(code(&jn, "NaN"), "unexpected");
}

#[test]
fn objects() {
    let j = make();
    for (src, want) in [
        ("{}", "{}"),
        (r#"{"a":1}"#, r#"{"a":1}"#),
        ("{a:1}", r#"{"a":1}"#),
        ("{a:1,b:2}", r#"{"a":1,"b":2}"#),
        ("{a:1, 'b':2}", r#"{"a":1,"b":2}"#),
        ("{ nested: { x: 1 } }", r#"{"nested":{"x":1}}"#),
        ("{$id:1, _n:2, a1:3}", r#"{"$id":1,"_n":2,"a1":3}"#),
    ] {
        assert_eq!(parsed(&j, src), want, "{src:?}");
    }
}

#[test]
fn arrays() {
    let j = make();
    for (src, want) in [
        ("[]", "[]"),
        ("[1]", "[1]"),
        ("[1,2,3]", "[1,2,3]"),
        (r#"["a","b"]"#, r#"["a","b"]"#),
        ("[[1,2],[3,4]]", "[[1,2],[3,4]]"),
    ] {
        assert_eq!(parsed(&j, src), want, "{src:?}");
    }
}

#[test]
fn trailing_commas() {
    let j = make();
    for (src, want) in [
        ("[1,2,3,]", "[1,2,3]"),
        ("{a:1,b:2,}", r#"{"a":1,"b":2}"#),
        ("[1,]", "[1]"),
        ("{a:1,}", r#"{"a":1}"#),
        ("[ 1 , 2 , ]", "[1,2]"),
    ] {
        assert_eq!(parsed(&j, src), want, "{src:?}");
    }
}

#[test]
fn comments() {
    let j = make();
    for (src, want) in [
        ("// hello\n42", "42"),
        ("/* block */ 42", "42"),
        ("{ a: 1, /* mid */ b: 2 }", r#"{"a":1,"b":2}"#),
        ("[/* a */ 1, /* b */ 2]", "[1,2]"),
        ("/* multi\nline\ncomment */ [1,2]", "[1,2]"),
    ] {
        assert_eq!(parsed(&j, src), want, "{src:?}");
    }

    // Hash comments are not JSON5 by default.
    assert_eq!(code(&j, "# nope\n42"), "unexpected");

    // Can be enabled.
    let jh = parser(|o| o.hash_comment = true);
    assert_eq!(parsed(&jh, "# hello\n42"), "42");
}

#[test]
fn strings() {
    let j = make();
    for (src, want) in [
        (r#""hello""#, "hello"),
        ("'hello'", "hello"),
        (r#""he said \"hi\"""#, r#"he said "hi""#),
        (r#"'he said \'hi\''"#, "he said 'hi'"),
        (r#""a\tb""#, "a\tb"),
        (r#""a\nb""#, "a\nb"),
        (r#""aAb""#, "aAb"),
        (r#""a\x41b""#, "aAb"),
        (r#""\0""#, "\0"),
        ("\"line1\\\nline2\"", "line1line2"),
        ("\"line1\\\r\nline2\"", "line1line2"),
    ] {
        match parse_with(&j, src) {
            Ok(tabnas::Value::String(got)) => assert_eq!(got, want, "{src:?}"),
            other => panic!("{src:?}: {other:?}"),
        }
    }

    // Backticks are not JSON5 by default.
    assert_eq!(code(&j, "`backtick`"), "unexpected");

    // Can be enabled.
    let jb = parser(|o| o.backtick_string = true);
    assert_eq!(parsed(&jb, "`backtick`"), r#""backtick""#);
}

#[test]
fn rejects_non_json5() {
    let j = make();
    for src in [
        "", "foo", "0o17", "0b101", "1_000", "1,2,3", "a:1", "{,}", "{10:1}",
    ] {
        assert!(parse_with(&j, src).is_err(), "{src:?} expected an error");
    }
}

/// Mirrors the TypeScript `require-value` test: empty input under the
/// default requireValue option fails with `json5_empty` (message matching
/// /JSON5/), while requireValue=false lets it parse to `null`.
#[test]
fn require_value() {
    let j = make();
    let error = parse_with(&j, "").expect_err("empty input is an error");
    assert_eq!(error.code, "json5_empty");
    assert!(error.to_string().contains("JSON5"), "{error}");
    assert_eq!(error.detail, "JSON5 input must contain a value");
    assert!(error.hint.contains("top-level value"), "{}", error.hint);

    let error = parse_with(&j, "   // nothing").expect_err("no value is an error");
    assert_eq!(error.code, "json5_no_value");

    // A direct `parse` on the instance also fails (lex.empty is off),
    // though with the engine's generic error; the package-level
    // `parse_with` is the counterpart of the TypeScript wrapped start
    // and the Go `Parse(j, src)`.
    assert!(j.parse("").is_err());

    // Allow empty input.
    let jopt = parser(|o| o.require_value = false);
    assert_eq!(parsed(&jopt, ""), "null");
    assert_eq!(parsed(&jopt, "  // nothing\n/* here */"), "null");
}

#[test]
fn non_strict_options() {
    let js = parser(|o| {
        o.octal = true;
        o.binary = true;
        o.number_separator = true;
    });
    assert_eq!(parsed(&js, "0o17"), "15");
    assert_eq!(parsed(&js, "0b101"), "5");
    assert_eq!(parsed(&js, "1_000"), "1000");

    let jnh = parser(|o| o.hex = false);
    assert_eq!(code(&jnh, "0x1F"), "unexpected");
}

#[test]
fn strict_value_toggle() {
    // With strictValue disabled, bare words parse as text strings
    // (jsonic's default text fallback).
    let j = parser(|o| o.strict_value = false);
    assert_eq!(parsed(&j, "foo"), r#""foo""#);
}

#[test]
fn json5_spec_example() {
    let j = make();
    let src = "{
      // comments
      unquoted: 'and you can quote me on that',
      singleQuotes: 'I can use \"double quotes\" here',
      lineBreaks: \"Look, Mom! \\
No \\\\n's!\",
      hexadecimal: 0xdecaf,
      leadingDecimalPoint: .8675309, andTrailing: 8675309.,
      positiveSign: +1,
      trailingComma: 'in objects', andIn: ['arrays',],
      \"backwardsCompatible\": \"with JSON\",
    }";
    assert_eq!(
        parsed(&j, src),
        concat!(
            r#"{"unquoted":"and you can quote me on that","#,
            r#""singleQuotes":"I can use \"double quotes\" here","#,
            r#""lineBreaks":"Look, Mom! No \\n's!","hexadecimal":912559,"#,
            r#""leadingDecimalPoint":0.8675309,"andTrailing":8675309,"positiveSign":1,"#,
            r#""trailingComma":"in objects","andIn":["arrays"],"backwardsCompatible":"with JSON"}"#
        )
    );
}

#[test]
fn json_is_json5() {
    let j = make();
    for src in [
        "{}",
        "[]",
        r#"{"a":1,"b":"two","c":null,"d":true,"e":false}"#,
        r#"[1,2.5,-3,1e10,"s",null,true,false]"#,
        r#"{"nested":{"list":[1,{"x":null}]}}"#,
    ] {
        // Compared in the fixture data model, where every number is an
        // f64 as in JSON.parse, so `1` and `1.0` are one value.
        let want = tabnas_support::parse_expect(src).expect("valid JSON");
        let got = to_value(&parse_with(&j, src).unwrap_or_else(|error| panic!("{src:?}: {error}")));
        assert!(
            tabnas_support::equal_value(&got, &want),
            "{src:?}: {}",
            tabnas_support::format_value(&got)
        );
    }
}

// --- Keys: decoded, not just validated (test/spec/keys.tsv pins the
// --- table; this pins the API detail that the decoded name is the key)

#[test]
fn unquoted_keys_are_decoded() {
    let j = make();
    let value = parse_with(&j, r"{sigΣma:1, A:2}").expect("it parses");
    let tabnas::Value::Object(map) = value else {
        panic!("not an object: {value:?}");
    };
    let keys: Vec<&str> = map.keys().map(String::as_str).collect();
    assert_eq!(keys, ["sigΣma", "A"]);
}

// --- The API surface -------------------------------------------------

#[test]
fn options_round_trip_through_the_plugin_bag() {
    let options = Json5Options {
        hash_comment: true,
        require_value: false,
        ..Default::default()
    };
    assert_eq!(Json5Options::from_value(&options.to_value()), options);

    // A key the bag does not carry keeps its default, as in Go.
    let partial = Json5Options::from_json(&serde_json::json!({ "octal": true }));
    assert_eq!(
        partial,
        Json5Options {
            octal: true,
            ..Default::default()
        }
    );
    // As does a key of the wrong type.
    let odd = Json5Options::from_json(&serde_json::json!({ "hex": "no" }));
    assert_eq!(odd, Json5Options::default());
}

#[test]
fn the_plugin_installs_through_use_plugin_and_records_its_options() {
    let mut parser = tabnas_jsonic::make();
    let overrides = Json5Options {
        backtick_string: true,
        ..Default::default()
    };
    parser
        .use_plugin(plugin(), Some(overrides.to_value()))
        .expect("the plugin installs");
    assert_eq!(parsed(&parser, "`x`"), r#""x""#);
    assert_eq!(code(&parser, ""), "json5_empty");

    // The engine's option bag carries the merged defaults.
    let bag = parser.plugin_options("json5").expect("the bag is recorded");
    assert_eq!(Json5Options::from_value(bag), overrides);
}

#[test]
fn the_plain_install_function_matches_the_plugin() {
    let mut direct = tabnas_jsonic::make();
    json5(&mut direct, &Json5Options::default()).expect("the plugin installs");
    let through_make = make();
    for src in ["{a:1,}", "[1,2,]", "0x10", "Infinity", "'s'"] {
        assert_eq!(parsed(&direct, src), parsed(&through_make, src), "{src:?}");
    }
    assert_eq!(code(&direct, "foo"), code(&through_make, "foo"));
}

#[test]
fn parse_with_on_a_parser_without_the_plugin_is_a_plain_parse() {
    let jsonic = tabnas_jsonic::make();
    assert_eq!(parsed(&jsonic, "a:1"), r#"{"a":1}"#);
    assert_eq!(
        json(&parse_with(&jsonic, "").expect("jsonic accepts empty")),
        "undefined"
    );
}

#[test]
fn errors_carry_the_position() {
    let j = make();
    let error = parse_with(&j, "[1,\n  foo]").expect_err("bare text is rejected");
    assert_eq!(error.code, "unexpected");
    assert_eq!((error.row, error.col), (2, 3));
}

#[test]
fn a_string_line_continuation_is_stripped_inside_strings_only() {
    let j = make();
    assert_eq!(parsed(&j, "['a\\\nb', \"p\\\r\nq\"]"), r#"["ab","pq"]"#);
    assert!(parse_with(&j, "[1,\\\n2]").is_err());
    // A backslash does NOT continue a line comment.
    assert_eq!(parsed(&j, "// c\\\n1"), "1");
}

// --- The shared default parser ----------------------------------------

#[test]
fn parse_uses_a_shared_default_and_is_safe_across_threads() {
    assert_eq!(
        json(&parse("{a:[1,2,],b:0x10}").unwrap()),
        r#"{"a":[1,2],"b":16}"#
    );
    assert_eq!(parse("").unwrap_err().code, "json5_empty");

    let sources: Arc<Vec<&'static str>> = Arc::new(vec![
        "{a:1}",
        "[1,2,3,]",
        "'x'",
        "Infinity",
        "// c\n42",
        "{sig\\u03A3ma:1}",
    ]);
    let handles: Vec<_> = (0..8)
        .map(|_| {
            let sources = Arc::clone(&sources);
            thread::spawn(move || {
                for _ in 0..50 {
                    for src in sources.iter() {
                        parse(src).unwrap_or_else(|error| panic!("{src:?}: {error}"));
                    }
                }
            })
        })
        .collect();
    for handle in handles {
        handle.join().expect("a parsing thread panicked");
    }
}
